### ISSUE TITLE
Make better examples

### ISSUE THREAD

**Issue by ballsteve** [ballsteve/xrust#30]
Make examples that are simple and elegant. There should be no examples that are embarrassing.

If examples are embarrassing, improve the API to make the example more elegant.

**Comment by ballsteve** [comment]
Add examples to iterators in Forest Node struct

**Comment by mdubinko** [comment]
Request: an example of using XPath to extract something simple, like a single text node, from a Forest.

**Comment by ballsteve** [comment]
Great idea, but the Forest module is in the process of being replaced by a (hopefully) better tree implementation that uses strong and weak references to maintain parent/child relationships.

Once that code is ready, I'll write up an example along the lines you suggest.

**Comment by mdubinko** [comment]
Since this section of the code is getting rewritten, I won't depend on this code. But as part of learning how it works today, I came up with this example of using XPath but not XSLT. Recording here for posterity. If there's a better place, let me know.

```
fn test_xrust_xpath() {
    let parsed_expr = xpath::parse("/child::a/child::*");
    assert!(parsed_expr.is_ok());
    let xpath = parsed_expr.expect("Error parsing expression");
    let mut f = Forest::new();
    let input_doc = f.grow_tree("<a><b>c</b></a>").expect("XML parse failed..."); 
    let doc_node = f.get_ref(input_doc).expect("tree doesn't exist!").get_doc_node();
    let output_doc = f.plant_tree();
    let evaluator = Evaluator::new();
    let result = evaluator.evaluate(
        Some(Sequence::from(doc_node)), // context
        Some(1), // position
        &xpath, // XPath
        &mut f, // forest
        input_doc, // source document
        output_doc); // result document
    assert!(result.is_ok());
    assert_eq!(result.expect("Evaluation error").to_xml(Some(&f)), "<b>c</b>");
}
```

**Comment by ballsteve** [comment]
Thanks for that. It covers the current code using the Forest module.

Do you have any suggestions for how the API can be improved?

For example, the code I'm developing now defines a "Node" trait for both the document and nodes (unlike Forest which has separate Document and Node objects). This trait could define an "evaluate XPath expression" method that sets the self object as the context for the expression. Something like:

```rust
let source_doc = XML::from("<a><b><c>d</c></b></a>")
    .expect("unable to parse XML");
// Start with the document root node and evaluate an absolute path
let result1: Sequence = source_doc
    .xpath("/child::a/child::*")
    .expect("unable to evaluate XPath");
// The context will be a node and evaluate a relative path
let result2: Sequence = result1
    .xpath("child::*")
    .expect("unable to evaluate XPath");
assert_eq!(result2.to_xml(), "<c>d</c>")
```

Thoughts?

**Comment by mdubinko** [comment]
I like this a lot.
It would be nice if the new Forest implementation could be used standalone, or along with an XPath (or XSLT/XQuery) module that adds greater capability.
There would be a flexible set of ways to serialize a Sequence too, right? (Here is where Canonical XML might come in.)
Another question I had about the earlier example: there should be some way to easily set up / modify the static context, right? In particular, the set of builtin functions available...

**Comment by ballsteve** [comment]
An example has been added to the dev branch, named 'issue-30'. It demonstrates the use of the new transformation engine to evaluate an XPath expression on a source XML document without the use of XSLT.

**Comment by ballsteve** [comment]
NB. right now it is in the branch 'new-features-2023-11-28'. This branch will soon be merged into the dev branch.
